### ISSUE TITLE
Can't remove Offered DLCs, can't reliably send funds from wallet state

### ISSUE THREAD

**Issue by user411** [bitcoin-s/bitcoin-s#3813]
**Describe the bug**
This is a set of behaviors I've seen on my primary wallet. I am having issues using 'Send' to reliably send funds and cannot 'Cancel DLC' on two DLCs in Offered state in the wallet. It feels like some preceding transaction never made it to the network or a utxo being "not reserved" is keeping additional processes from happening.

I tried to send a number of transactions from this wallet to my secondary wallet and one of many tries broadcast and transmitted funds.

wallet-ts GetFundedAddresses() returns exactly one utxo at this point.
```
{
  address: 'tb1qhls7qarxudmktzkdsky9jc84nffyn6gyhmef5e',
  value: 11960234
}
```

Trying to delete a Offered DLC:
```
[info] 2021-11-11T15:43:52UTC INFO [Server] Unhandled error in server:
[info] java.lang.IllegalArgumentException: requirement failed: Some utxos are not reserved, got Vector(SegwitV0SpendingInfo(TransactionOutPoint(b671d0b5ac844a33934909da909b7fd8c6f39abdc9ee14920bd2fe39c5ffa4f9:1),TransactionOutput(124792 sats,wpkh(556cce341a0ef7120a9a84ee6ba617b083ad33a8)),m/84'/1'/0'/1/102,P2WPKHWitnessV0(List(0209eee74b5527f046e2cf09c318c814e78717a4aa40669ff8f800b5c35919ba4a, )),DoubleSha256BDigestBE(b671d0b5ac844a33934909da909b7fd8c6f39abdc9ee14920bd2fe39c5ffa4f9),BroadcastSpent,Some(DoubleSha256BDigestBE(669f0012ab800caf4212800ca4ffc00f8a189c45513f1a59b2684a1214612ec8)),Some(26)))
[info] 	at scala.Predef$.require(Predef.scala:337)
[info] 	at org.bitcoins.wallet.internal.UtxoHandling.unmarkUTXOsAsReserved(UtxoHandling.scala:338)
[info] 	at org.bitcoins.wallet.internal.UtxoHandling.unmarkUTXOsAsReserved$(UtxoHandling.scala:335)
[info] 	at org.bitcoins.wallet.Wallet.unmarkUTXOsAsReserved(Wallet.scala:47)
[info] 	at org.bitcoins.dlc.wallet.DLCWallet.$anonfun$cancelDLC$6(DLCWallet.scala:268)
[info] 	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:434)
[info] 	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
[info] 	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
[info] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[info] 	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
[info] 	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
[info] 	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
[info] 	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
[info] 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
[info] 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
[info] 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
[info] 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
[info] 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

**To Reproduce**
1. Load wallet backup
2. Try to Cancel the Offered DLCs / try to Send

The wallet backup is 86mb and I can't attach it here. Please contact me for "2021-1111 macbook bitcoin-s-backup.zip".

**Comment by user411** [comment]
Triying

`./bitcoin-s-cli rescan --force`

Logged

```
[info] 2021-11-11T16:36:49UTC INFO [DLCWallet$DLCWalletImpl] Starting rescanning the wallet from None to None
[info] 2021-11-11T16:37:08UTC ERROR [SafeDatabase] Error when executing query delete [delete from "addresses" where "addresses"."address" in ('tb1qucv84ux5k4y4wvtcpn5g0d05r7x0c48fnlsjy2')]
[info] 2021-11-11T16:37:08UTC ERROR [SafeDatabase] Error when executing query delete [delete from "addresses" where "addresses"."address" in ('tb1q24az0kmahmaq4zxmfg2t849d4ds2vs4p3q92xq')]
[info] 2021-11-11T16:37:08UTC ERROR [SafeDatabase] org.sqlite.SQLiteException: [SQLITE_CONSTRAINT_FOREIGNKEY]  A foreign key constraint failed (FOREIGN KEY constraint failed)
[info] 2021-11-11T16:37:08UTC ERROR [SafeDatabase] org.sqlite.SQLiteException: [SQLITE_CONSTRAINT_FOREIGNKEY]  A foreign key constraint failed (FOREIGN KEY constraint failed)
[info] 2021-11-11T16:37:08UTC INFO [DLCWallet$DLCWalletImpl] Finished rescanning the wallet. It took 19020ms
```

Found these addreses still had labels
```
GetAddressLabels() tb1q24az0kmahmaq4zxmfg2t849d4ds2vs4p3q92xq
labels { result: [ 'test' ], error: null }
DropAddressLabels() tb1q24az0kmahmaq4zxmfg2t849d4ds2vs4p3q92xq
GetAddressLabels() tb1qucv84ux5k4y4wvtcpn5g0d05r7x0c48fnlsjy2
labels { result: [ 'test' ], error: null }
DropAddressLabels() tb1qucv84ux5k4y4wvtcpn5g0d05r7x0c48fnlsjy2
```

Nuked them...

**Comment by user411** [comment]
Rescanned again. Trying to delete the remaining Offered DLC I get an error

```
[info] java.lang.IllegalArgumentException: requirement failed: Some utxos are not reserved, got Vector(SegwitV0SpendingInfo(TransactionOutPoint(b79506ef9ad3649d89bba73a4ac8d5cb8346170a55377e57a43da109c1df8e09:0),TransactionOutput(12490 sats,wpkh(3097e3d292b5da16135dbc8d85ec8548bb47ab0d)),m/84'/1'/0'/0/24,P2WPKHWitnessV0(List(038098bd33f89bb23276a9ecda37b2e36cdef9b51d88c829cc3c2890a8ade51db0, )),DoubleSha256BDigestBE(b79506ef9ad3649d89bba73a4ac8d5cb8346170a55377e57a43da109c1df8e09),ConfirmedReceived,None,Some(84)), SegwitV0SpendingInfo(TransactionOutPoint(2cc7aa1f90807e479f09de102a291291d3ffd0302be7dc170896dff0fd9f5f0c:0),TransactionOutput(2289 sats,wpkh(eb7ebb31193ac77da91aec9f88231aaf76ba338f)),m/84'/1'/0'/1/38,P2WPKHWitnessV0(List(03e7827801d5372134b3f084cc0cd0e5d7819842ba2d0fca78f364bf53c5842b2b, )),DoubleSha256BDigestBE(2cc7aa1f90807e479f09de102a291291d3ffd0302be7dc170896dff0fd9f5f0c),ConfirmedReceived,None,Some(85)), SegwitV0SpendingInfo(TransactionOutPoint(3333ab3137c8eec617cde50d5d493e47f60d8188670776dc662bb12d1620e0ae:0),TransactionOutput(14881 sats,wpkh(79c4d4ff33934d78a0658070fa6014b5fcaf5bbc)),m/84'/1'/0'/0/77,P2WPKHWitnessV0(List(028ba1b62a32b31792debaf862eb8323455fb9083bcb02674734da1372e08b3386, )),DoubleSha256BDigestBE(3333ab3137c8eec617cde50d5d493e47f60d8188670776dc662bb12d1620e0ae),ConfirmedReceived,None,Some(86)), SegwitV0SpendingInfo(TransactionOutPoint(0ad58de249d0d2a8b819d8f28b3b623f44d128e135d732e8beef93b53110e679:0),TransactionOutput(9000 sats,wpkh(7b15460361c455e3052f98f2daeb143cd0e30676)),m/84'/1'/0'/0/4,P2WPKHWitnessV0(List(0286722a0c13981da8b8bddf2b1dc307576a82d5aae3b5bf1334826aee5582a20c, )),DoubleSha256BDigestBE(0ad58de249d0d2a8b819d8f28b3b623f44d128e135d732e8beef93b53110e679),ConfirmedReceived,None,Some(90)), SegwitV0SpendingInfo(TransactionOutPoint(97983ba550ac85b024f76c94324980fb37cff6c5e1855db9da799df97ddfa5b2:0),TransactionOutput(85000 sats,wpkh(a69be48e96bc4f90a47d52f338b6a4ba56db9f39)),m/84'/1'/0'/0/2,P2WPKHWitnessV0(List(033618e58d0f4dd45b0dbe6cc79661c595a1c7a978e22be9b364a383713038cb47, )),DoubleSha256BDigestBE(97983ba550ac85b024f76c94324980fb37cff6c5e1855db9da799df97ddfa5b2),ConfirmedReceived,None,Some(93)))
```

**Comment by Christewart** [comment]
Closing this for now, let's reopen if it isn'tfully solved by #3817 / #3816

**Comment by user411** [comment]
Going back to old wallet to test changes made. Resynced master, rebuilt cli and bundle, added old seed file and dlcdb / walletdb. Started up `bundle/run` did a `./bitcoin-s-cli rescan --force` and got

```
[info] 2021-11-15T17:30:50UTC INFO [DLCWallet$DLCWalletImpl] Starting rescanning the wallet from None to None
[info] 2021-11-15T17:31:09UTC ERROR [SafeDatabase] Error when executing query delete [delete from "addresses" where "addresses"."address" in ('tb1qucv84ux5k4y4wvtcpn5g0d05r7x0c48fnlsjy2')]
[info] 2021-11-15T17:31:09UTC ERROR [SafeDatabase] Error when executing query delete [delete from "addresses" where "addresses"."address" in ('tb1q24az0kmahmaq4zxmfg2t849d4ds2vs4p3q92xq')]
[info] 2021-11-15T17:31:09UTC ERROR [SafeDatabase] org.sqlite.SQLiteException: [SQLITE_CONSTRAINT_FOREIGNKEY]  A foreign key constraint failed (FOREIGN KEY constraint failed)
[info] 2021-11-15T17:31:09UTC ERROR [SafeDatabase] org.sqlite.SQLiteException: [SQLITE_CONSTRAINT_FOREIGNKEY]  A foreign key constraint failed (FOREIGN KEY constraint failed)
[info] 2021-11-15T17:31:09UTC INFO [DLCWallet$DLCWalletImpl] Finished rescanning the wallet. It took 19023ms
```

Haven't tried to cancel any DLCs or move funds on the wallet.

When I first restored the wallet, I added the seed file only and not the backed up dlc/wallet dbs. This gave me a positive wallet balance and no dlcs. Turning the server off, adding the dlc/wallet dbs and rescanning got me to the errors here and 0 balances.

For the second wallet with issues, I resynced master, rebuilt cli and bundle, added old seed file and dlcdb / walletdb. Started up `bundle/run` did a `./bitcoin-s-cli rescan --force` and got

```
[info] 2021-11-15T17:48:27UTC INFO [DLCWallet$DLCWalletImpl] Starting rescanning the wallet from None to None
[info] 2021-11-15T17:48:29UTC ERROR [SafeDatabase] Error when executing query delete [delete from "addresses" where "addresses"."address" in ('tb1quweu6c9h3fw5k9xfpt4r0g02dfs2h7r3uu9jdr')]
[info] 2021-11-15T17:48:29UTC ERROR [SafeDatabase] org.sqlite.SQLiteException: [SQLITE_CONSTRAINT_FOREIGNKEY]  A foreign key constraint failed (FOREIGN KEY constraint failed)
[info] 2021-11-15T17:48:29UTC INFO [DLCWallet$DLCWalletImpl] Finished rescanning the wallet. It took 1701ms
```
